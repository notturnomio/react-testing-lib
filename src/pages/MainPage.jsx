import { useEffect, useState } from 'react';
import Counter from '../components/Counter/Counter';
// import '../App.css';

function MainPage() {
  const [data, setData] = useState(null);
  const [toggle, setToggle] = useState(false);
  const [inputValue, setInputValue] = useState('');

  const onClick = () => setToggle((prev) => !prev);

  useEffect(() => {
    setTimeout(() => {
      setData({});
    }, 500);
  }, []);

  return (
    <div className='App' data-testid='main-page'>
      <h2>Main Page</h2>
      {data && (
        <div style={{ color: 'red' }} data-testid='async-text'>
          My async text line
        </div>
      )}
      <h1 className='App-header' data-testid='value-elm'>
        {inputValue}
      </h1>
      <div>
        <input
          type='text'
          placeholder='input value...'
          onChange={(e) => setInputValue(e.target.value)}
        />
      </div>
      <div style={{ padding: '20px 0' }}>
        <button data-testid='toggle-btn' onClick={onClick}>
          Toggle hidden text
        </button>
      </div>
      {toggle === true && (
        <div data-testid='toggle-elm' className='App-header'>
          Toggle
        </div>
      )}
      <Counter />
    </div>
  );
}

export default MainPage;
