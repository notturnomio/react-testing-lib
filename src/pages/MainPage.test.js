/* eslint-disable testing-library/prefer-presence-queries */
/* eslint-disable testing-library/no-debugging-utils */
import { fireEvent, render, screen } from '@testing-library/react';
import MainPage from './MainPage';

describe('TEST MainPage', () => {
  test('checks sample page elements', () => {
    render(<MainPage url='/' />);
    const helloWorldElement = screen.getByText(/hello world/i);
    const btn = screen.getByRole('button');
    const input = screen.getByPlaceholderText(/input value/i);
    expect(helloWorldElement).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
    // screen.debug();
  });

  test('tests no element', () => {
    render(<MainPage url='/' />);
    const hello2 = screen.queryByText(/hello2/i);
    expect(hello2).toBeNull();
  });

  test('tests async element', async () => {
    render(<MainPage url='/' />);
    // act(() => {
    //   jest.advanceTimersByTime(1000);
    // });
    const hello3 = await screen.findByTestId('async-text');
    expect(hello3).toBeInTheDocument();
    expect(hello3).toHaveStyle({ color: 'red' });
  });

  test('tests click', () => {
    render(<MainPage url='/' />);
    const toggleBtn = screen.getByTestId('toggle-btn');
    expect(screen.queryByTestId('toggle-elm')).not.toBeInTheDocument();
    fireEvent.click(toggleBtn);
    expect(screen.getByTestId('toggle-elm')).toBeInTheDocument();
    fireEvent.click(toggleBtn);
    expect(screen.queryByTestId('toggle-elm')).not.toBeInTheDocument();
  });

  test('tests input', () => {
    render(<MainPage url='/' />);
    const input = screen.getByPlaceholderText(/input value/i);
    expect(screen.queryByTestId('value-elm')).toContainHTML('');
    fireEvent.input(input, {
      target: { value: '123' },
    });
    expect(screen.queryByTestId('value-elm')).toContainHTML('123');
    fireEvent.input(input, {
      target: { value: '' },
    });
    fireEvent.input(input, {
      target: { value: '12345' },
    });
    expect(screen.queryByTestId('value-elm')).toContainHTML('12345');
  });
});
