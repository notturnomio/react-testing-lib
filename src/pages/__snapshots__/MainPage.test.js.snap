// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST MainPage checks sample page elements 1`] = `
<input
  placeholder="input value..."
  type="text"
/>
`;
